### ISSUE TITLE
tesseract threads never more than 2

### ISSUE THREAD

**Issue by marlemion** [jbarlow83/OCRmyPDF#135]
Hi, I am using ocrmypdf with --jobs 4, but the number of simultaneous tesseract thread never goes beyond 2. My system has 5 cores, so I would like to youse at least for of them. gs processes are spawned in parallel up to 4.

Is this an intended limitation? tesseract is 4alpha.

**Comment by jbarlow83** [comment]
Yes, it's intended because someone reported severely slowdowns on their
Linux box with an i7, because both ocrmypdf and tesseract 4 try to use all
cores.

I do know that clang-compiled Tesseract seems to be single threaded even
when compiled with OpenMP. Also if running in docker the container may
impose limits.

Does tesseract 4 use all cores when run on its own?

Can you tell me more about your environment?

**Comment by marlemion** [comment]
I use arch on a virtual machine on a server, whichs architecture I am not really familiar with (AMD something, I can ask if you are interested). 24GB RAM. tesseract obviously does not multithread. ocrmypdf spawns maximum two tesseract threads, each using up to one core 100%. This machine is merely for text processing and file indexing. There is an instance of Datafari (solr, manifolcf, tika combination) running on it as well. This software also uses tesseract. I limited it to ocr images only (not pdf, this is done by python script calling ocrmypdf). So it happens that more than 5 instances are running in parallel (two from ocrmypdf the other three by datafari).

Indexing can be controlled, so I could separate the ocrmypdf scanning and index scanning timely. This will be the default behavior anyway: first ocr, then index. So I could use some extra power for ocrmypdf. We have +350000 files from which at least 10% are PDFs without text. This means, half cores, double time. It already took me upt to 10 day halfway through.

Is this somehow helpful for you? I'd test, if you tell me how.

**Comment by jbarlow83** [comment]
>tesseract obviously does not multithread

Ah! Let's get that solved first.

Tesseract 4 *should* now multithread if the LSTM engine is working correctly. Check that you can run ``tesseract --oem 1`` and have it work properly, it should say it using the LSTM engine and it should run threads sometimes. I don't think the old engine has been retrofitted to multithread, or it may do so too briefly to notice.

If Tesseract 4 is trying to work with Tesseract 3's tessdata, ``--oem 1`` will exit with an error, but ``--oem 3`` will work because Tesseract 4 falls back to the Tesseract 3 engine.

**Comment by jbarlow83** [comment]
For Tesseract 4 with ocrmypdf I recommend ``--output-type pdf --pdf-renderer tess4`` because Ghostscript (the default --output-type) ruins the OCR text. I'm working on this bug on the tesseract side.

**Comment by marlemion** [comment]
Hi. Thanky for your reply. I am using the alpha version released on the github site of tesseract:

https://github.com/tesseract-ocr/tesseract/releases

So no real git head. It works veery nicely together with ocrmypdf, no text corruptions or the like. As I use it in a working environment right know I would not rather not touch it right now. Especially as you say that there are problems with text corruption in later versions.

Hence, I get:

```
tesseract --oem 1 [...]
read_params_file: parameter not found: ����

tesseract --oem 3 [...]
read_params_file: parameter not found: ����

tesseract -v
tesseract 4.00.00alpha
 leptonica-1.74.1
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.5.1) : libpng 1.6.28 : libtiff 4.0.7 : zlib 1.2.11 : libwebp 0.5.2 : libopenjp2 2.1.2

ocrmypdf --output-type pdf --pdf-renderer tess4 [...]
   INFO - Tesseract v4.x.alpha found. OCRmyPDF support is experimental.
  ERROR - --pdf-renderer tess4 requires Tesseract 4.x commit 3d9fb3b or later
```
Does it mean I am still using more or less tesseract 3.x? tessdata is from here:

https://github.com/tesseract-ocr/tessdata/releases (4.00)

**Comment by jbarlow83** [comment]
The text corruption issue is caused by Ghostscript which ocrmypdf uses under certain settings. The default settings of ocrmypdf are safe. I am *changing* my earlier recommendation; use the defaults for ``--output-type`` and ``--pdf-renderer`` by *omitting* these parameters, since they are safer  than using them (although they will transcode images in some cases).

``parameter not found`` seems like a tesseract command line syntax error.
You should be able to run exactly this command using Tesseract 4's testing/phototest.tif as an example:

```
$ tesseract --oem 1 testing/phototest.tif stdout
Page 1
This is a lot of 12 point text to test the
ocr code and see if it works on all types
of file format.

The quick brown dog jumped over the
lazy fox. The quick brown dog jumped
over the lazy fox. The quick brown dog
jumped over the lazy fox. The quick
brown dog jumped over the lazy fox.
```

and get error-free output. Then find out if Tess4 is using multiple threads.

If you like Ubuntu, there is a PPA that auto-updates with compiled new Tesseract releases: ``ppa:alex-p/tesseract-ocr``.

**Comment by marlemion** [comment]
Hm, it doesn't have this option:

```
tesseract --oem 1 test.jpg stdout
read_params_file: parameter not found: ����

tesseract -h
Usage:
  tesseract --help | --help-psm | --version
  tesseract --list-langs [--tessdata-dir PATH]
  tesseract --print-parameters [options...] [configfile...]
  tesseract imagename|stdin outputbase|stdout [options...] [configfile...]

OCR options:
  --tessdata-dir PATH   Specify the location of tessdata path.
  --user-words PATH     Specify the location of user words file.
  --user-patterns PATH  Specify the location of user patterns file.
  -l LANG[+LANG]        Specify language(s) used for OCR.
  -c VAR=VALUE          Set value for config variables.
                        Multiple -c arguments are allowed.
  -psm NUM              Specify page segmentation mode.
NOTE: These options must occur before any configfile.

Page segmentation modes:
  0    Orientation and script detection (OSD) only.
  1    Automatic page segmentation with OSD.
  2    Automatic page segmentation, but no OSD, or OCR.
  3    Fully automatic page segmentation, but no OSD. (Default)
  4    Assume a single column of text of variable sizes.
  5    Assume a single uniform block of vertically aligned text.
  6    Assume a single uniform block of text.
  7    Treat the image as a single text line.
  8    Treat the image as a single word.
  9    Treat the image as a single word in a circle.
 10    Treat the image as a single character.
 11    Sparse text. Find as much text as possible in no particular order.
 12    Sparse text with OSD.
 13    Raw line. Treat the image as a single text line,
                        bypassing hacks that are Tesseract-specific.

Single options:
  -h, --help            Show this help message.
  --help-psm            Show page segmentation modes.
  -v, --version         Show version information.
  --list-langs          List available languages for tesseract engine.
  --print-parameters    Print tesseract parameters to stdout.

tesseract -l deu test.jpg stdout
30.4.1912 Einweihungsfeier
„Von großer Bedeutung für die hiesigen Schulverhältnisse ist die Vollendung des neuen Schulhauses.[...]
```

**Comment by marlemion** [comment]
For now I set the limit in pipeline.py from 2 to 4 and it seems to work very well.

**Comment by jacotec** [comment]
Hi,

I have a similar issue, using a ESXi VM for OCRing documents I put in my Alfresco. Although I've changed my calling script to never process more than 2 documents at once, ocrmypdf fires as many tesseract threads as my VM has cores assigned when it processes a multipage document.

Even after searching for hours and playing around with cpulimit -c 2 I did not find a working solution to limit ocrmypdf to use only one or two cores per call.

What is this "--jobs" parmeter mentioned, and what's this limit in "pipeline.py"? --jobs is not documented in the ocrmypdf docs ... is there any way to prevent ocrmypdf from using all cores?

**Comment by jacotec** [comment]
OK ... I found it ;-)

I was searching in the wrong ocrmypdf documentation in the internet, found the "jobs" thing here in the readme.

Calling it with "ocrmypdf --jobs 2 ..." successfully limits the number of threads to 2 per job.

Sorry for the inconvenience ...

**Comment by jbarlow83** [comment]
Glad you found what you needed. 

The "pipeline.py" limit the other user referenced is a hardcoded rule that ocrmypdf will launch no more than two Tesseract 4 processes, because Tesseract 4 is multithreaded on its own when compiled correctly. If you're not using Tesseract 4 this is not something to worry about. I might expose this option in a future release.

**Comment by Gugols** [comment]
Will post here since it is relevant to the `--jobs` flag:
For example, if I have 8 CPU cores and I want to use 6 of them fully, does it make sense to use `--jobs 6` flag? Since it seems that only 1-2 of cores are used to full potential and the rest are under low load.
<img width="984" alt="Screenshot 2019-04-23 at 22 38 38" src="https://user-images.githubusercontent.com/7780981/56610858-2b689300-6619-11e9-9eac-44ff1e4113a3.png">

>The "pipeline.py" limit the other user referenced is a hardcoded rule that ocrmypdf will launch no more than two Tesseract 4 processe

**Comment by jbarlow83** [comment]
> The "pipeline.py" limit the other user referenced is a hardcoded rule that ocrmypdf will launch no more than two Tesseract 4 processe

The current version of ocrmypdf does not limit Tesseract 4 processes, but you didn't say what version you have. If you observed low CPU on a recent version please open a new issue and provided all the details in the issue template.

Some stages of the pipeline are single threaded (not Tesseract).
